### PR TITLE
Migrate CPU_tensor_apply to TensorIterator in aten/src/ATen/native/TensorCompare.cpp

### PR DESCRIPTION
**Pull request by ifedan**
https://github.com/pytorch/pytorch/issues/24497
https://github.com/pytorch/pytorch/issues/24498

### CODE REVIEW

File: aten/src/ATen/native/TensorCompare.cpp [hunk 1]
```diff
@@ -14,31 +16,26 @@ void where_cpu(
...
-          self,
-          other,
-          [](scalar_t& ret_val,
-             const bool& cond_val,
-             const scalar_t& self_val,
-             const scalar_t& other_val) {
-            ret_val = cond_val ? self_val : other_val;
-          });
-    }
+    at::native::cpu_serial_kernel(
```
Comment (fmassa) [inline aten/src/ATen/native/TensorCompare.cpp]: question: why do we need serial execution here?

File: aten/src/ATen/native/TensorCompare.cpp [hunk 2]
```diff
@@ -6,6 +6,8 @@
 #include <ATen/native/ReduceOpsUtils.h>
 #include <c10/util/Exception.h>
 #include <ATen/native/cpu/TensorCompareKernel.h>
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/cpu/vec256/vec256.h>
```
Comment (fmassa) [inline aten/src/ATen/native/TensorCompare.cpp]: is this import necessary?

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

This can be probably further optimized by using `cpu_kernel_vec` using `Vec256::blendv`, but that can be done in a follow-up PR.

**Review by VitalyFedyunin — APPROVED** [review]
LGTM but you will likely get better performance by moving into cpu subfolder and allow compiler vectorization.

Also it would be nice to see performance results.

File: aten/src/ATen/native/TensorCompare.cpp (hunk 2 above)
Comment (ifedan) [reply aten/src/ATen/native/TensorCompare.cpp]: I removed this, and change implementation to parallel

**Comment by ifedan** [comment]
@VitalyFedyunin: Correct me if I wrong, but vectorization requires that all vectorized arguments have same size. In my case I have uint8/bool and scalar_t

**Comment by VitalyFedyunin** [comment]
Actually you right, they are different in size and vectorization quite impractical here. We can skip /cpu/ part here.

**Pull request closed without merging**